staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,700 @@
+				klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, immediate", rd, fs.Name, fs.Spec.DistinguisherMethod, plName)
+				return fs.Name, fs.Spec.DistinguisherMethod, plName, nil
+			}
+			plState.numPending++
+			matchID := &plName
+			klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID)
+			startCalls := 0
+			startFn := func(ctx context.Context, hashValue uint64) (execute bool, afterExecution func()) {
+				startCalls++
+				klog.V(7).Infof("For matchID=%p, startFn(ctx, %v) startCalls:=%d", matchID, hashValue, startCalls)
+				if startCalls > 1 {
+					panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p startCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, startCalls))
+				}
+				req := func() fq.Request {
+					cfgCtl.lock.Lock()
+					defer cfgCtl.lock.Unlock()
+					plState.numPending--
+					req, idle1 := plState.queues.StartRequest(ctx, hashValue, rd.RequestInfo, rd.User)
+					if idle1 {
+						cfgCtl.maybeReapLocked(plName, plState)
+					}
+					return req
+				}()
+				if req == nil {
+					klog.V(7).Infof("For matchID=%p, startFn(ctx, %v): reject", matchID, hashValue)
+					return false, func() {}
+				}
+				exec, idle2, afterExec := req.Wait()
+				execID := &exec
+				if idle2 {
+					cfgCtl.maybeReap(plName)
+				}
+				klog.V(7).Infof("For matchID=%p, startFn(..) => exec=%v, execID=%p", matchID, exec, execID)
+				afterCalls := 0
+				return exec, func() {
+					afterCalls++
+					klog.V(7).Infof("For matchID=%p, execID=%p, after() calls:=%d", matchID, execID, afterCalls)
+					if afterCalls > 1 {
+						panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p, execID=%p, afterCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, execID, afterCalls))
+					}
+					idle3 := afterExec()
+					if idle3 {
+						cfgCtl.maybeReap(plName)
+					}
+				}
+			}
+			return fs.Name, fs.Spec.DistinguisherMethod, plName, startFn
+		}
+	}
+	// This can never happen because every configState has a
+	// FlowSchema that matches everything
+	panic(rd)
+}
+
+// Call this after getting a clue that the given priority level is undesired and idle
+func (cfgCtl *configController) maybeReap(plName string) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	plState := cfgCtl.priorityLevelStates[plName]
+	if plState != nil && (plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle()) {
+		klog.V(3).Infof("Triggered API Priority and Fairness config reload because priority level %s became idle", plName)
+		cfgCtl.configQueue.Add(0)
+	}
+}
+
+// Call this if both (1) plState.queues is non-nil and reported being
+// idle, and (2) cfgCtl's lock has not been released since then.
+func (cfgCtl *configController) maybeReapLocked(plName string, plState *priorityLevelState) {
```

Comment:
nit: not really sure but probably `maybeReapUnlocked` is a better name?

Reply:
This file is following the same naming convention as the queueset type: suffix "Locked" means the caller has to hold the lock.  I will add a comment in this file.

Reply:
Added comment, on the configController type.